File: platypus-js-web-client/src/main/java/com/eas/widgets/BorderPane.java
package com.eas.widgets;

import com.eas.core.HasPublished;
import com.eas.core.XElement;
import com.eas.menu.HasComponentPopupMenu;
import com.eas.menu.PlatypusPopupMenu;
import com.eas.ui.HasEventsExecutor;
import com.eas.ui.HasJsFacade;
import com.eas.ui.HorizontalPosition;
import com.eas.ui.VerticalPosition;
import com.eas.ui.events.AddEvent;
import com.eas.ui.events.AddHandler;
import com.eas.ui.events.EventsExecutor;
import com.eas.ui.events.HasAddHandlers;
import com.eas.ui.events.HasHideHandlers;
import com.eas.ui.events.HasRemoveHandlers;
import com.eas.ui.events.HasShowHandlers;
import com.eas.ui.events.HideEvent;
import com.eas.ui.events.HideHandler;
import com.eas.ui.events.RemoveEvent;
import com.eas.ui.events.RemoveHandler;
import com.eas.ui.events.ShowEvent;
import com.eas.ui.events.ShowHandler;
import com.eas.widgets.containers.BorderPanel;
import com.google.gwt.core.client.JavaScriptObject;
import com.google.gwt.dom.client.Style;
import com.google.gwt.event.dom.client.ContextMenuEvent;
import com.google.gwt.event.dom.client.ContextMenuHandler;
import com.google.gwt.event.logical.shared.HasResizeHandlers;
import com.google.gwt.event.logical.shared.ResizeEvent;
import com.google.gwt.event.logical.shared.ResizeHandler;
import com.google.gwt.event.shared.HandlerRegistration;
import com.google.gwt.user.client.ui.HasEnabled;
import com.google.gwt.user.client.ui.Widget;

public class BorderPane extends BorderPanel implements HasJsFacade, HasEnabled, HasComponentPopupMenu, HasEventsExecutor, HasShowHandlers, HasHideHandlers, HasResizeHandlers, HasAddHandlers,
        HasRemoveHandlers, HasChildrenPosition {

	protected EventsExecutor eventsExecutor;
	protected PlatypusPopupMenu menu;
	protected boolean enabled = true;
	protected String name;
	protected JavaScriptObject published;

	public BorderPane() {
		super();
		getElement().getStyle().setOverflow(Style.Overflow.HIDDEN);
		getElement().getStyle().setPosition(Style.Position.RELATIVE);
	}

	public BorderPane(int aVGap, int aHGap) {
		this();
		getElement().getStyle().setOverflow(Style.Overflow.HIDDEN);
		getElement().getStyle().setPosition(Style.Position.RELATIVE);
		setHgap(aHGap);
		setVgap(aVGap);
	}

	@Override
	public HandlerRegistration addAddHandler(AddHandler handler) {
		return addHandler(handler, AddEvent.getType());
	}

	@Override
	public HandlerRegistration addRemoveHandler(RemoveHandler handler) {
		return addHandler(handler, RemoveEvent.getType());
	}

	@Override
	public HandlerRegistration addResizeHandler(ResizeHandler handler) {
		return addHandler(handler, ResizeEvent.getType());
	}

	@Override
	public void onResize() {
		super.onResize();
		if (isAttached()) {
			ResizeEvent.fire(this, getElement().getOffsetWidth(), getElement().getOffsetHeight());
		}
	}

	@Override
	public HandlerRegistration addHideHandler(HideHandler handler) {
		return addHandler(handler, HideEvent.getType());
	}

	@Override
	public HandlerRegistration addShowHandler(ShowHandler handler) {
		return addHandler(handler, ShowEvent.getType());
	}

	@Override
	public void setVisible(boolean visible) {
		boolean oldValue = isVisible();
		super.setVisible(visible);
		if (oldValue != visible) {
			if (visible) {
				ShowEvent.fire(this, this);
			} else {
				HideEvent.fire(this, this);
			}
		}
	}

	@Override
	public EventsExecutor getEventsExecutor() {
		return eventsExecutor;
	}

	@Override
	public void setEventsExecutor(EventsExecutor aExecutor) {
		eventsExecutor = aExecutor;
	}

	@Override
	public PlatypusPopupMenu getPlatypusPopupMenu() {
		return menu;
	}

	protected HandlerRegistration menuTriggerReg;

	@Override
	public void setPlatypusPopupMenu(PlatypusPopupMenu aMenu) {
		if (menu != aMenu) {
			if (menuTriggerReg != null)
				menuTriggerReg.removeHandler();
			menu = aMenu;
			if (menu != null) {
				menuTriggerReg = super.addDomHandler(new ContextMenuHandler() {

					@Override
					public void onContextMenu(ContextMenuEvent event) {
						event.preventDefault();
						event.stopPropagation();
						menu.setPopupPosition(event.getNativeEvent().getClientX(), event.getNativeEvent().getClientY());
						menu.show();
					}
				}, ContextMenuEvent.getType());
			}
		}
	}

	@Override
	public boolean isEnabled() {
		return enabled;
	}

	@Override
	public void setEnabled(boolean aValue) {
		boolean oldValue = enabled;
		enabled = aValue;
		if (!oldValue && enabled) {
			getElement().<XElement> cast().unmask();
		} else if (oldValue && !enabled) {
			getElement().<XElement> cast().disabledMask();
		}
	}

	@Override
	public String getJsName() {
		return name;
	}

	@Override
	public void setJsName(String aValue) {
		name = aValue;
	}

	@Override
	public boolean remove(Widget w) {
		boolean res = super.remove(w);
		if (res)
			RemoveEvent.fire(this, w);
		return res;
	}

	public void add(Widget aWidget, int aPlace, int aSize) {
		switch (aPlace) {
		case HorizontalPosition.LEFT:
			setLeftComponent(aWidget, aSize);
			break;
		case HorizontalPosition.RIGHT:
			setRightComponent(aWidget, aSize);
			break;
		case VerticalPosition.TOP:
			setTopComponent(aWidget, aSize);
			break;
		case VerticalPosition.BOTTOM:
			setBottomComponent(aWidget, aSize);
			break;
		default:
			setCenterComponent(aWidget);
			break;
		}
	}

	public Widget getLeftComponent() {
		for (Widget w : getChildren()) {
			Direction d = getResolvedDirection(getWidgetDirection(w));
			if (d == Direction.WEST) {
				return w;
			}
		}
		return null;
	}

	public void setLeftComponent(Widget w, double size) {
		Widget old = getLeftComponent();
		if (old != w) {
			if (old != null)
				remove(old);
			if (w != null) {
				Widget centerWidget = super.getCenter();
				if (centerWidget != null)
					super.remove(centerWidget);
				addWest(w, size);
				if (centerWidget != null)
					super.add(centerWidget);
				AddEvent.fire(this, w);
			}
		}
	}

	public Widget getRightComponent() {
		for (Widget w : getChildren()) {
			Direction d = getResolvedDirection(getWidgetDirection(w));
			if (d == Direction.EAST) {
				return w;
			}
		}
		return null;
	}

	public void setRightComponent(Widget w, double size) {
		Widget old = getRightComponent();
		if (old != w) {
			if (old != null)
				remove(old);
			if (w != null) {
				Widget centerWidget = super.getCenter();
				if (centerWidget != null)
					super.remove(centerWidget);
				addEast(w, size);
				if (centerWidget != null)
					super.add(centerWidget);
				AddEvent.fire(this, w);
			}
		}
	}

	public Widget getTopComponent() {
		for (Widget w : getChildren()) {
			Direction d = getResolvedDirection(getWidgetDirection(w));
			if (d == Direction.NORTH) {
				return w;
			}
		}
		return null;
	}

	public void setTopComponent(Widget w, double size) {
		Widget old = getTopComponent();
		if (old != w) {
			if (old != null)
				remove(old);
			if (w != null) {
				Widget centerWidget = super.getCenter();
				if (centerWidget != null)
					super.remove(centerWidget);
				addNorth(w, size);
				if (centerWidget != null)
					super.add(centerWidget);
				AddEvent.fire(this, w);
			}
		}
	}

	public Widget getBottomComponent() {
		for (Widget w : getChildren()) {
			Direction d = getResolvedDirection(getWidgetDirection(w));
			if (d == Direction.SOUTH) {
				return w;
			}
		}
		return null;
	}

	public void setBottomComponent(Widget w, double size) {
		Widget old = getBottomComponent();
		if (old != w) {
			if (old != null)
				remove(old);
			if (w != null) {
				Widget centerWidget = super.getCenter();
				if (centerWidget != null)
					super.remove(centerWidget);
				addSouth(w, size);
				if (centerWidget != null)
					super.add(centerWidget);
				AddEvent.fire(this, w);
			}
		}
	}

	public Widget getCenterComponent() {
		return getCenter();
	}

	public void setCenterComponent(Widget w) {
		Widget old = getCenter();
		if (old != w) {
			if (old != null)
				remove(old);
			if (w != null) {
				super.add(w);
				AddEvent.fire(this, w);
			}
		}
	}

	@Override
	public JavaScriptObject getPublished() {
		return published;
	}

	@Override
	public void setPublished(JavaScriptObject aValue) {
		if (published != aValue) {
			published = aValue;
			if (published != null) {
				publish(this, aValue);
			}
		}
	}

	private native static void publish(HasPublished aWidget, JavaScriptObject published)/*-{
		var VerticalPosition = @com.eas.ui.JsUi::VerticalPosition;		
		var HorizontalPosition = @com.eas.ui.JsUi::HorizontalPosition;		
		Object.defineProperty(published, "hgap", {
			get : function(){
				return aWidget.@com.eas.widgets.BorderPane::getHgap()();
			},
			set : function(aValue){
				aWidget.@com.eas.widgets.BorderPane::setHgap(I)(aValue);
			}
		});
		Object.defineProperty(published, "vgap", {
			get : function(){
				return aWidget.@com.eas.widgets.BorderPane::getVgap()();
			},
			set : function(aValue){
				aWidget.@com.eas.widgets.BorderPane::setVgap(I)(aValue);
			}
		});
		Object.defineProperty(published, "leftComponent", {
			get : function() {
				var comp = aWidget.@com.eas.widgets.BorderPane::getLeftComponent()();
				return @com.eas.core.Utils::checkPublishedComponent(Ljava/lang/Object;)(comp);
			},
			set : function(aChild) {
				aWidget.@com.eas.widgets.BorderPane::setLeftComponent(Lcom/google/gwt/user/client/ui/Widget;D)(aChild.unwrap(), toAdd.width);
			}
		});
		Object.defineProperty(published, "rightComponent", {
			get : function() {
				var comp = aWidget.@com.eas.widgets.BorderPane::getRightComponent()();
				return @com.eas.core.Utils::checkPublishedComponent(Ljava/lang/Object;)(comp);
			},
			set : function(aChild) {
				aWidget.@com.eas.widgets.BorderPane::setRightComponent(Lcom/google/gwt/user/client/ui/Widget;D)(aChild.unwrap(), toAdd.width);
			}
		});
		Object.defineProperty(published, "topComponent", {
			get : function() {
				var comp = aWidget.@com.eas.widgets.BorderPane::getTopComponent()();
				return @com.eas.core.Utils::checkPublishedComponent(Ljava/lang/Object;)(comp);
			},
			set : function(aChild) {
				aWidget.@com.eas.widgets.BorderPane::setTopComponent(Lcom/google/gwt/user/client/ui/Widget;D)(aChild.unwrap(), toAdd.height);
			}
		});
		Object.defineProperty(published, "bottomComponent", {
			get : function() {
				var comp = aWidget.@com.eas.widgets.BorderPane::getBottomComponent()();
				return @com.eas.core.Utils::checkPublishedComponent(Ljava/lang/Object;)(comp);
			},
			set : function(aChild) {
				aWidget.@com.eas.widgets.BorderPane::setBottomComponent(Lcom/google/gwt/user/client/ui/Widget;D)(aChild.unwrap(), toAdd.height);
			}
		});
		Object.defineProperty(published, "centerComponent", {
			get : function() {
				var comp = aWidget.@com.eas.widgets.BorderPane::getCenterComponent()();
				return @com.eas.core.Utils::checkPublishedComponent(Ljava/lang/Object;)(comp);
			},
			set : function(aChild) {
			 	aWidget.@com.eas.widgets.BorderPane::setCenterComponent(Lcom/google/gwt/user/client/ui/Widget;)(aChild.unwrap());
			}
		});
		published.add = function(toAdd, region, aSize) {
			if(toAdd != undefined && toAdd != null && toAdd.unwrap != undefined){
				if(toAdd.parent == published)
					throw 'A widget already added to this container';
				if(!region){
					region = VerticalPosition.CENTER;
				}
				switch (region) {
					case VerticalPosition.CENTER:
						aWidget.@com.eas.widgets.BorderPane::setCenterComponent(Lcom/google/gwt/user/client/ui/Widget;)(toAdd.unwrap());
						break;  
					case VerticalPosition.TOP: 
						if (!aSize) {
							aSize = toAdd.height;
							if (!aSize) {
								aSize = 32;
							}
						}
						aWidget.@com.eas.widgets.BorderPane::setTopComponent(Lcom/google/gwt/user/client/ui/Widget;D)(toAdd.unwrap(), aSize);
						break;  
					case VerticalPosition.BOTTOM: 
						if (!aSize) {
							aSize = toAdd.height;
							if (!aSize) {
								aSize = 32;
							}
						}
						aWidget.@com.eas.widgets.BorderPane::setBottomComponent(Lcom/google/gwt/user/client/ui/Widget;D)(toAdd.unwrap(), aSize);
						break;  
					case HorizontalPosition.LEFT: 
						if (!aSize) {
							aSize = toAdd.width;
							if (!aSize) {
								aSize = 32
							}
						}
						aWidget.@com.eas.widgets.BorderPane::setLeftComponent(Lcom/google/gwt/user/client/ui/Widget;D)(toAdd.unwrap(), aSize);
						break;  
					case HorizontalPosition.RIGHT: 
						if (!aSize) {
							aSize = toAdd.width;
							if (!aSize) {
								aSize = 32
							}
						}
						aWidget.@com.eas.widgets.BorderPane::setRightComponent(Lcom/google/gwt/user/client/ui/Widget;D)(toAdd.unwrap(), aSize);
						break;  
				}
			}
		}
	}-*/;

	@Override
	public int getTop(Widget aWidget) {
		assert aWidget.getParent() == this : "widget should be a child of this container";
		return aWidget.getElement().getParentElement().getOffsetTop();
	}

	@Override
	public int getLeft(Widget aWidget) {
		assert aWidget.getParent() == this : "widget should be a child of this container";
		return aWidget.getElement().getParentElement().getOffsetLeft();
	}
}

File: platypus-js-web-client/src/app/forms/border-pane.js
define([
    '../extend',
    '../ui',
    './container'
], function (
        extend,
        Ui,
        Container) {
    function BorderPane(hgap, vgap) {
        Container.call(this);

        var self = this;

        var center;
        var left;
        var right;
        var top;
        var bottom;
        var bottomHeight;
        var topHeight;
        var leftWidth;
        var rightWidth;

        var hgap = 0;
        var vgap = 0;

        Object.defineProperty(this, 'hgap', {
            get: function () {
                return hgap;
            },
            set: function (aValue) {
                hgap = aValue;
                recalcCenterMargins();
            }
        });

        Object.defineProperty(this, 'vgap', {
            get: function () {
                return vgap;
            },
            set: function (aValue) {
                vgap = aValue;
                recalcCenterMargins();
            }
        });

        var superAdd = this.add;
        function add(w, aPlace) {
            if (arguments.lenth < 2) {
                self.centerComponent = w;
            } else {
                switch (aPlace) {
                    case Ui.HorizontalPosition.LEFT:
                        self.leftComponent = w;
                        break;
                    case Ui.HorizontalPosition.RIGHT:
                        self.rightComponent = w;
                        break;
                    case Ui.VerticalPosition.TOP:
                        self.topComponent = w;
                        break;
                    case Ui.VerticalPosition.BOTTOM:
                        self.bottomComponent = w;
                        break;
                    default:
                        self.centerComponent = w;
                        break;
                }
            }
        }
        Object.defineProperty(this, 'add', {
            get: function () {
                return add;
            }
        });

        Object.defineProperty(this, 'leftComponent', {
            get: function () {
                return left;
            },
            set: function (w) {
                if (w !== left) {
                    if (left) {
                        superRemove(left);
                    }
                    left = w;
                    if (left) {
                        leftWidth = left.width;
                        if (!leftWidth)
                            leftWidth = 30;
                        formatLeft();
                        superAdd(left);
                    }
                }
            }
        });
        Object.defineProperty(this, 'rightComponent', {
            get: function () {
                return right;
            },
            set: function (w) {
                if (w !== right) {
                    if (right) {
                        superRemove(right);
                    }
                    right = w;
                    if (right) {
                        rightWidth = right.width;
                        if (!rightWidth)
                            rightWidth = 30;
                        formatRight();
                        superAdd(right);
                    }
                }
            }
        });
        Object.defineProperty(this, 'topComponent', {
            get: function () {
                return top;
            },
            set: function (w) {
                if (w !== top) {
                    if (top) {
                        superRemove(top);
                    }
                    top = w;
                    if (top) {
                        topHeight = top.height;
                        if (!topHeight)
                            topHeight = 30;
                        formatTop();
                        superAdd(top);
                    }
                }
            }
        });
        Object.defineProperty(this, 'bottomComponent', {
            get: function () {
                return bottom;
            },
            set: function (w) {
                if (w !== bottom) {
                    if (bottom) {
                        superRemove(bottom);
                    }
                    bottom = w;
                    if (bottom) {
                        bottomHeight = bottom.height;
                        if (!bottomHeight)
                            bottomHeight = 30;
                        formatBottom();
                        superAdd(bottom);
                    }
                }
            }
        });
        Object.defineProperty(this, 'centerComponent', {
            get: function () {
                return center;
            },
            set: function (w) {
                if (w !== center) {
                    if (center) {
                        superRemove(center);
                    }
                    center = w;
                    if (center) {
                        centerHeight = center.height;
                        formatCenter();
                        recalcCenterMargins();
                        superAdd(center);
                    }
                }
            }
        });

        function formatLeft() {
            if (left) {
                var ls = left.element.style;
                ls.position = 'absolute';
                ls.overflow = 'hidden';
                ls.left = 0 + 'px';
                ls.width = leftWidth + 'px';
                ls.top = topHeight + 'px';
                ls.bottom = bottomHeight + 'px';
            }
        }

        function formatRight() {
            if (right) {
                var rs = right.element.style;
                rs.position = 'absolute';
                rs.pverflow = 'hidden';
                rs.right = 0 + 'px';
                rs.width = rightWidth + 'px';
                rs.top = topHeight + 'px';
                rs.bottom = bottomHeight + 'px';
            }
        }

        function formatTop() {
            if (top) {
                var ts = top.element.style;
                ts.position = 'absolute';
                ts.overflow = 'hidden';
                ts.top = 0 + 'px';
                ts.height = topHeight + 'px';
                ts.left = leftWidth + 'px';
                ts.right = rightWidth + 'px';
            }
        }

        function formatBottom() {
            if (bottom) {
                var bs = bottom.element.style;
                bs.position = 'absolute';
                bs.overflow = 'hidden';
                bs.bottom = 0 + 'px';
                bs.height = bottomHeight + 'px';
                bs.left = leftWidth + 'px';
                bs.right = rightWidth + 'px';
            }
        }

        function formatCenter() {
            if (center) {
                var cs = center.element.style;
                cs.position = 'absolute';
                cs.overflow = 'hidden';
                cs.bottom = bottomHeight + 'px';
                cs.top = topHeight + 'px';
                cs.left = leftWidth + 'px';
                cs.right = rightWidth + 'px';
            }
        }

        function checkParts(w) {
            if (left === w) {
                left = null;
                leftWidth = 0;
            }
            if (right === w) {
                right = null;
                rightWidth = 0;
            }
            if (top === w) {
                top = null;
                topHeight = 0;
            }
            if (bottom === w) {
                bottom = null;
                bottomHeight = 0;
            }
            if (center === w) {
                center = null;
            }
        }

        var superRemove = this.remove;
        function remove(widgetOrIndex) {
            var removed = superRemove(widgetOrIndex);
            checkParts(removed);
            return removed;
        }

        function recalcCenterMargins() {
            if (center) {
                var cs = center.element.style;
                cs.marginLeft = hgap + 'px';
                cs.marginRight = hgap + 'px';
                cs.marginTop = vgap + 'px';
                cs.marginBottom = vgap + 'px';
            }
        }

        function ajustWidth(w, width) {
            if (left === w) {
                leftWidth = width;
                formatLeft();
                formatTop();
                formatBottom();
                formatCenter();
            } else if (right === w) {
                rightWidth = width;
                formatRight();
                formatTop();
                formatBottom();
                formatCenter();
            }
        }
        Object.defineProperty(this, 'ajustWidth', {
            get: function () {
                return ajustWidth;
            }
        });

        function ajustHeight(w, height) {
            if (top === w) {
                topHeight = height;
                formatTop();
                formatLeft();
                formatRight();
                formatCenter();
            } else if (bottom === w) {
                bottomHeight = height;
                formatBottom();
                formatLeft();
                formatRight();
                formatCenter();
            }
        }
        Object.defineProperty(this, 'ajustHeight', {
            get: function () {
                return ajustHeight;
            }
        });

        function getTop(w) {
            if (w.parent !== this)
                throw "Widget should be a child of this container";
            return w.element.offsetTop;
        }
        Object.defineProperty(this, 'getTop', {
            get: function () {
                return getTop;
            }
        });

        function getLeft(w) {
            if (w.parent !== this)
                throw "Widget should be a child of this container";
            return w.element.getOffsetLeft();
        }
        Object.defineProperty(this, 'getLeft', {
            get: function () {
                return getLeft;
            }
        });
    }
    extend(BorderPane, Container);
    return BorderPane;
});